### PR TITLE
Make WindowState bindable

### LINKED ISSUE #990: Make WindowState bindable
Also requires `WindowStateChanged` in `IWindowImpl`

### PR DESCRIPTION
**Pull request by jp2masa**
Fixes #990 
## Changes
- Added `WindowStateChanged` property to `IWindowImpl`, and implemented it for Win32, MonoMac and Gtk3 (only tested for Win32)
- Added `WindowStateProperty` to `Window`

### CODE REVIEW

**Comment by kekekeks** [comment]
You can test (partially, without X11-related shenanigans) GTK3 backend on Windows.

https://mega.nz/#!ZtwBBDwQ!j9iuS6sC7e1as2Csj5wlV3Bw1szXqWDDIA-y1tQuuEM

Unpack somewhere and add it to `PATH`. The add `UseGtk3` after `UsePlatformDetect` in `ControlCatalog.Desktop`

File: src/Gtk/Avalonia.Gtk3/WindowImpl.cs [hunk 1]
```diff
@@ -9,8 +9,27 @@ namespace Avalonia.Gtk3
...
         public WindowImpl() : base(Native.GtkWindowNew(GtkWindowType.TopLevel))
         {
+            ConnectEvent("window-state-event", OnWindowStateChanged);
+        }
+
+        private unsafe bool OnWindowStateChanged(IntPtr gtkwidget, IntPtr ev, IntPtr userdata)
+        {
+            var windowStateEvent = (GdkEventWindowState*)ev;
+            var newWindowState = windowStateEvent->new_window_state;
+            var windowState = newWindowState.HasFlag(GdkWindowState.Maximized) ? WindowState.Maximized
```
Comment (kekekeks) [inline src/Gtk/Avalonia.Gtk3/WindowImpl.cs]: Iconified should take precedence over Maximized

Comment (jp2masa) [reply src/Gtk/Avalonia.Gtk3/WindowImpl.cs]: `IsMiniaturized` in MonoMac should take precendence over `IsZoomed` too?

Comment (kekekeks) [reply src/Gtk/Avalonia.Gtk3/WindowImpl.cs]: Yes, I guess so. It doesn't matter if window is maximized or not, if it's not visible to the user anyway.

I'm also considering to change the API and make those two separate properties.

**Comment by jp2masa** [comment]
The GTK3 backend on Windows doesn't work for me. I can see the application icon in the taskbar, but no window is shown.

**Comment by milkinllc** [comment]
I'm having the same issue as @jp2masa but when I add up UseSkia() I am able to see working application content but without window system decorations (unable to move or resize window:). The same problem with system decorations occurs when running net core version. With net core version UseSkia() configuration call is unnecessary (works as per @kekekeks instructions).

**Comment by grokys** [comment]
@jp2masa just wondering what the status of this is? Were there problems? It's not clear from the comments what the problem was. Does this PR break GTK3?

**Comment by jp2masa** [comment]
I only tested this on Windows.

**Comment by grokys** [comment]
So the reason this hasn't been merged is that it's only been tested on Windows?

**Comment by jp2masa** [comment]
I think so, I tried testing GTK3 on Windows using @kekekeks instructions, but it didn't work, I'll try again later today.

**Comment by Sorien** [comment]
in Gtk you should probably use https://developer.gnome.org/gtk3/stable/GtkWidget.html#GtkWidget-window-state-event

**Comment by Sorien** [comment]
GTk looks ok if you rebase i can test it tomorrow

**Comment by grokys** [comment]
@Sorien did you get chance to try this out?

**Comment by Sorien** [comment]
@grokys just tested at latest ubuntu and its working without any problem.

**Review by grokys — APPROVED** [review]
Looks good to me! Thanks @jp2masa and sorry about the (massive) delay.

**Pull request merged**